From af07c4f68c0708033c8304f0185a908450b5f94e Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sun, 10 Aug 2014 18:20:00 -0400
Subject: [iommu] amd: Fix resource leak in iommu_init_device()

Message-id: <20140810182000.24055.54163.stgit@gir.stowe>
Patchwork-id: 87576
O-Subject: [RHEL7.1 PATCH 05/19] iommu/amd: Fix resource leak in iommu_init_device()
Bugzilla: 1127467
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Dutile <ddutile@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1127467
Upstream Status: e644a013fe67f2bccd54378b88556d07fa2714d6

commit e644a013fe67f2bccd54378b88556d07fa2714d6
Author: Radmila Kompova <xkompo00@stud.fit.vutbr.cz>
Date:   Thu May 2 17:24:25 2013 +0200

    iommu/amd: Fix resource leak in iommu_init_device()

    Detected by cppcheck.

    Signed-off-by: Kamil Dudka <kdudka@redhat.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/amd_iommu.c b/drivers/iommu/amd_iommu.c
index 6dc6594..72531f0 100644
--- a/drivers/iommu/amd_iommu.c
+++ b/drivers/iommu/amd_iommu.c
@@ -456,8 +456,10 @@ static int iommu_init_device(struct device *dev)
  }
 
  ret = init_iommu_group(dev);
- if (ret)
+ if (ret) {
+  free_dev_data(dev_data);
   return ret;
+ }
 
  if (pci_iommuv2_capable(pdev)) {
   struct amd_iommu *iommu;
-- 
1.7.1